pkg/loader/fileloader.go

```diff
@@ -155,110 +160,99 @@ func (l *fileLoader) New(path string) (ifc.Loader, error) {
 	}
 	if isRepoUrl(path) {
-		// This works well enough for purpose at hand to detect
-		// previously visited URLs and thus avoid cycles.
-		if err := isPathEqualToOrAbove(path, l.roots); err != nil {
+		// Avoid cycles.
+		if err := l.errIfPreviouslySeenUri(path); err != nil {
 			return nil, err
 		}
-		return newGitLoader(path, l.fSys, l.roots, l.cloner)
+		return newGitLoader(path, l.fSys, l.referrer, l.cloner)
 	}
 	if filepath.IsAbs(path) {
 		return nil, fmt.Errorf("new root '%s' cannot be absolute", path)
 	}
 	return newFileLoaderAt(
-		filepath.Join(l.Root(), path), l.fSys, l.roots, l.cloner)
+		l.root.Join(path), l.fSys, l, l.cloner)
 }
 
 // newGitLoader returns a new Loader pinned to a temporary
 // directory holding a cloned git repo.
 func newGitLoader(
-	root string, fSys fs.FileSystem,
-	roots []string, cloner gitCloner) (ifc.Loader, error) {
-	tmpDirForRepo, pathInRepo, err := cloner(root)
+	uri string, fSys fs.FileSystem,
+	referrer *fileLoader, cloner gitCloner) (ifc.Loader, error) {
+	tmpDirForRepo, pathInRepo, err := cloner(uri)
 	if err != nil {
 		return nil, err
 	}
-	trueRoot := filepath.Join(tmpDirForRepo, pathInRepo)
-	if !fSys.IsDir(trueRoot) {
+	root, f, err := fSys.CleanedAbs(
+		filepath.Join(tmpDirForRepo, pathInRepo))
+	if err != nil {
+		return nil, err
+	}
+	if f != "" {
 		return nil, fmt.Errorf(
-			"something wrong cloning '%s'; unable to find '%s'",
-			root, trueRoot)
+			"'%s' refers to file '%s'; expecting directory", pathInRepo, f)
 	}
 	return &fileLoader{
-		roots:   append(roots, root, trueRoot),
-		fSys:    fSys,
-		cloner:  cloner,
-		cleaner: func() error { return fSys.RemoveAll(tmpDirForRepo) },
+		root:     root,
+		referrer: referrer,
+		uri:      uri,
+		fSys:     fSys,
+		cloner:   cloner,
+		cleaner:  func() error { return fSys.RemoveAll(tmpDirForRepo) },
 	}, nil
 }
 
-// isPathEqualToOrAbove tests whether the 1st argument,
-// viewed as a path to a directory, is equal to or above
-// any of the paths in the 2nd argument.  It's assumed
-// that all paths are cleaned, delinkified and absolute.
-func isPathEqualToOrAbove(path string, roots []string) error {
-	terminated := path + string(filepath.Separator)
-	for _, r := range roots {
-		if r == path || strings.HasPrefix(r, terminated) {
-			return fmt.Errorf(
-				"cycle detected: new root '%s' contains previous root '%s'",
-				path, r)
-		}
+// errIfArgEqualOrHigher tests whether the argument,
+// is equal to or above the root of any ancestor.
+func (l *fileLoader) errIfArgEqualOrHigher(
+	candidateRoot fs.ConfirmedDir) error {
+	if l.root.HasPrefix(candidateRoot) {
+		return fmt.Errorf(
+			"cycle detected: candidate root '%s' contains visited root '%s'",
+			candidateRoot, l.root)
 	}
-	return nil
+	if l.referrer == nil {
+		return nil
+	}
+	return l.referrer.errIfArgEqualOrHigher(candidateRoot)
+}
+
+func (l *fileLoader) errIfPreviouslySeenUri(uri string) error {
+	if strings.HasPrefix(l.uri, uri) {
```

Comment:
I'm wondering if we allow following cases here
```
github.com/project/repo.git/mydir?ref=master
github.com/project/repo.git/mydir?ref=test
```
These two URIs are different, but they point to the same dir.

Reply:
good point, added TODO.

I'd like to put gitUri on a firmer basis as its own type (not a string anymore) to make this kind of check saner.